File: 206-Reverse-Linked-List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null) return head;
        if (head.next == null) return head;

        ListNode curr = head;
        ListNode prev = null;
        ListNode temp = null;

        while (curr != null && curr.next != null) {
            temp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;
        }

        curr.next = prev;

        return curr;
    }
}

File: 206-Reverse-Linked-List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
private:
    ListNode* reverse_helper(ListNode* curr, ListNode* prev) {
        if (curr == nullptr) {
            return prev;
        }
        
        ListNode* temp = curr;
        curr = curr->next;
        temp->next = prev;
        prev = temp;
        return reverse_helper(curr, prev);
    }

public:
    ListNode* reverseList(ListNode* head) {
        // recursive solution
        // ListNode* prev = nullptr;
        // ListNode* curr = head;
        
        // head = reverse_helper(curr, prev);

        // return head;

        // iterative solution
        if (!head || !head->next) return head;

        ListNode* curr = head;
        ListNode* prev = nullptr;

        while(curr) {
            ListNode* temp = curr;
            curr = curr->next;
            temp->next = prev;
            prev = temp;
        }

        head = prev;
        return head;
    }

};